### PR TITLE
Hwpss filter

### PR DESCRIPTION
**Pull request by keskitalo**
Add HWPSS filter operator

### CODE REVIEW

**Comment by keskitalo** [comment]
And HWPSS filtering in the `FilterBin` operator.

File: src/toast/_libtoast/tod_filter.cpp [hunk 1]
```diff
@@ -125,6 +125,48 @@ void init_tod_filter(py::module & m) {
             None.
     )");
 
+    m.def("fourier",
+          [](py::buffer angle, py::buffer templates, size_t start_order,
+             size_t stop_order) {
+              pybuffer_check_1D <double> (angle);
+              py::buffer_info info_angle = angle.request();
+              py::buffer_info info_templates = templates.request();
```
Comment (tskisner) [inline src/toast/_libtoast/tod_filter.cpp:133]: I know we need to refactor the existing compiled code to move things directly into the `_libtoast` extension...  For consideration when writing new functions, there is a new templated helper function for extracting the raw pointers from a  `py::buffer` while simultaneously checking optional shape constraints.  See for example:  https://github.com/hpc4cmb/toast/blob/3252edf59d75a76adbc93f6947c1ab1bbb6fa3f9/src/toast/_libtoast/ops_mapmaker_utils.cpp#L120-L170

So this function would look like:
```
std::vector <int64_t> temp_shape(2);

double * pangle = extract_buffer <double> (angle, "angle", 1, temp_shape, {-1});
size_t nsample = temp_shape[0];

double * ptemplates = extract_buffer <double> (templates, "templates", 2, temp_shape, {-1, nsample});
size_t ntemplate = temp_shape[0];

if (ntemplate != 2 * (stop_order - start_order)) {
...
```
And the `extract_buffer` function raises an error if the shape constraints of the input buffer are not met.

**Review by tskisner — APPROVED** [review]
Some optional changes / suggestions, but no show-stoppers.  I know there will be conflicts with the other PR, but I'll address those when I rebase that branch after this is merged.

File: src/toast/ops/hwpfilter.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+    trend_order = Int(
+        5, help="Order of a Legendre polynomial to fit along with the HWPSS template."
+    )
+
+    filter_order = Int(
+        5, help="Order of a Fourier expansion to fit as a function of HWP angle."
+    )
+
+    view = Unicode(
```
[the remark this answers is not in the document]
Comment (keskitalo) [reply src/toast/ops/hwpfilter.py:108]: Fixed

File: src/toast/ops/hwpfilter.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        return check
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    @function_timer
+    def build_templates(self, obs):
+        """Construct the local HWPSS template hierarchy"""
+
+        views = obs.view[self.view]
```
[the remark this answers is not in the document]
Comment (keskitalo) [reply src/toast/ops/hwpfilter.py:151]: Removed.

File: src/toast/_libtoast/tod_filter.cpp (hunk 1 above)
Comment (keskitalo) [reply src/toast/_libtoast/tod_filter.cpp]: I agree, that is more elegant

**Pull request merged**